clang/include/clang/CIR/Dialect/IR/CIROps.td

```diff
@@ -4900,6 +4900,12 @@ def ClearCacheOp : CIR_Op<"clear_cache", [AllTypesMatch<["begin", "end"]>]> {
   }];
 }
 
+def InitListCtor : CIR_Op<"initlist.ctor"> {
+  let summary = "Starts a variable argument list";
+  let arguments = (ins CIR_PointerType:$initlist,
```

Comment:
I think it'd be better if the argument was just the arg list, and the op produced a result of the appropriate type. That way you can add the `SameTypeOperands` trait to verify that all the operand types are the same. You'll also want to verify that the result type is consistent with the argument type; `LoadOp` uses the `TypesMatchWith` trait for that, but our logic is a bit more complex, so you might need to implement a custom verifier (https://mlir.llvm.org/docs/DefiningDialects/Operations/#custom-verifier-code).

Reply:
If we do like this, we need to treat `std::initializer_list` as scalar instead of aggregate. I don't think it is a good way just for simplify the verify CIR.

Reply:
Let's push this discussion further and materialize our conclusion on it into the doc of this new op.

> If we do like this, we need to treat `std::initializer_list` as scalar instead of aggregate. I don't think it is a good way just for simplify the verify CIR.

From my understanding, "as scalar instead of aggregate" means treating `initializer_list` in high-level CIR as an value, or an pointer to value.

Let me humbly summarize both of your opinions here:

* A new pure SSA value for `initializer_list`, which probably introduces several new op with no extraordinary benefits, is of course not what we want.
* But we can *fuse* `cir.alloca` and current `cir.initializer_list.ctor` op into `cir.initializer_list`, which has result type `!cir.ptr<!tyInitList>`. That way we may get cleaner argument list, without evident IR design issue.
  * This benefit not only the verification, but also assembly format and possibly passes that touch the high-level op.
  * But I'm not sure that implicit `alloca` semantic won't cause any issue. Assumptions about memory operations should be taken care of.

Basically I think @HerrCai0907's current design is nice enough. Given our goal is to capture the list of values in `initializer_list`, only abstracting the constructor of it sounds reasonable to me. @smeenai Wdyt?

Reply:
My original idea was that `std::initializer_list` is just a struct, and we can treat that struct as an SSA value. Clang doesn't usually do so because it always performs ABI lowering, but we perform ABI lowering as a separate transform. Taking https://godbolt.org/z/dz88MrfWW as an example, note the lines:

```
      %17 = cir.load %2 : !cir.ptr<!ty_std3A3Ainitializer_list3Cint3E>, !ty_std3A3Ainitializer_list3Cint3E
      cir.call @_ZNSt6vectorIiEC1ESt16initializer_listIiE(%1, %17) : (!cir.ptr<!ty_std3A3Avector3Cint3E>, !ty_std3A3Ainitializer_list3Cint3E) -> ()
```

`%17` is an SSA value of type `std::initializer_list<int>` (which is just a `cir.struct` type with a particular format) being passed by value to a function, and I was thinking this op could behave similarly. Basically, right now the op usage looks like (in pseudo-CIR):

```
%init.alloc = cir.alloca std::initializer_list<int>
cir.initializer_list.ctor %init.alloc(1, 2, 3)
// somewhere later
%init.load = cir.load %init.alloc
cir.call some_function(%init.load)
```

We could instead theoretically have:

```
%init = cir.initializer_list.ctor 1, 2, 3
// somewhere later
cir.call some_function(%init)
```

We'd end up with the same result after lowering prepare, but the CIR before that would look nicer IMO. I haven't thought through all the implications of that though (in particular for cases like https://godbolt.org/z/PsK7d8PrG, which unfortunately runs into an assertion failure right now but where we'd still need the alloca), and I'm still pretty new to the project myself, so I might be missing something obvious :) I'm also fine with the current design.

Reply:
the process of codegen is like this, we check the expression type, if it a scalar, we want a op which return a value. if it an aggregate type, we ensure the memory and init it according to ptr.

It looks weird to do
```
if type.isRecord:
    if type.name == std::initializer_list:
        return scalar
    else:
         return aggregate
```

Reply:
Got it, your implementation is the way to go then.